File: Algorithm/C++/1000+/1013. Pairs of Songs With Total Durations Divisible by 60.cc
/*
 * Created on Wed Mar 20 2019 21:19:49
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// O(n) time, simple solution

#include <vector>

class Solution {
  public:
    int numPairsDivisibleBy60(std::vector<int>& time) {
      std::vector<int> count(60, 0);
      int result = 0;
      for (auto t : time) {
        result += count[(60 - (t%60)) % 60];
        count[t % 60]++;
      }
      return result;
    }
};

// faster solution
class Solution {
  public:
    int numPairsDivisibleBy60(std::vector<int>& time) {
      std::vector<int> count(60, 0);
      for (auto t : time) count[t % 60]++;
      int result = 0;
      for (int i = 0; i <= 30; ++i) {
        if (i == 0 || i == 30)
          result += (count[i]*(count[i]-1))/2;
        else
          result += count[i]*count[60-i];
      }
      return result;
    }
};

File: Algorithm/Python/1000+/1013. Pairs of Songs With Total Durations Divisible by 60.py
# -*- coding: utf-8 -*-
# Created on Wed Mar 20 2019 21:8:13
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# O(n) time, hashmap


class Solution(object):
    def numPairsDivisibleBy60(self, time):
        """
        :type time: List[int]
        :rtype: int
        """
        result, count = 0, [0]*60
        for t in time:
            result += count[(60-(t % 60)) % 60]
            count[t % 60] += 1
        return result


# faster solution

class Solution(object):
    def numPairsDivisibleBy60(self, time):
        """
        :type time: List[int]
        :rtype: int
        """
        count = [0] * 60
        for t in time:
            count[t % 60] += 1
        result = 0
        for i in xrange(31):
            if i == 0 or i == 30:
                result += (count[i] * (count[i] - 1)) >> 1
            else:
                result += count[i] * count[60 - i]
        return result
